tooling/lsp/src/requests/hover.rs

```diff
@@ -361,6 +369,144 @@ fn format_parent_module_from_module_id(
+                    self.gather_type_links(generic);
+                }
+            }
+            Type::Alias(type_alias, generics) => {
+                self.gather_type_alias_links(type_alias);
+                for generic in generics {
+                    self.gather_type_links(generic);
+                }
+            }
+            Type::TypeVariable(var, _) => {
+                self.gather_type_variable_links(var);
+            }
+            Type::TraitAsType(trait_id, _, generics) => {
+                let some_trait = self.interner.get_trait(*trait_id);
+                self.gather_trait_links(some_trait);
+                for generic in generics {
+                    self.gather_type_links(generic);
+                }
+            }
+            Type::NamedGeneric(var, _, _) => {
+                self.gather_type_variable_links(var);
+            }
+            Type::Function(args, return_type, env) => {
+                for arg in args {
+                    self.gather_type_links(arg);
+                }
+                self.gather_type_links(return_type);
+                self.gather_type_links(env);
+            }
+            Type::MutableReference(typ) => self.gather_type_links(typ),
+            Type::FieldElement
+            | Type::Integer(..)
+            | Type::Bool
+            | Type::String(_)
+            | Type::FmtString(_, _)
+            | Type::Unit
+            | Type::Forall(_, _)
+            | Type::Constant(_)
+            | Type::Quoted(_)
+            | Type::Error => (),
+        }
+    }
+
+    fn gather_struct_type_links(&mut self, struct_type: &Shared<StructType>) {
+        let struct_type = struct_type.borrow();
+        if let Some(lsp_location) =
+            to_lsp_location(self.files, struct_type.location.file, struct_type.name.span())
+        {
+            self.push_link(format_link(struct_type.name.to_string(), lsp_location));
+        }
+    }
+
+    fn gather_type_alias_links(&mut self, type_alias: &Shared<TypeAlias>) {
+        let type_alias = type_alias.borrow();
+        if let Some(lsp_location) =
+            to_lsp_location(self.files, type_alias.location.file, type_alias.name.span())
+        {
+            self.push_link(format_link(type_alias.name.to_string(), lsp_location));
+        }
+    }
+
+    fn gather_trait_links(&mut self, some_trait: &Trait) {
+        if let Some(lsp_location) =
+            to_lsp_location(self.files, some_trait.location.file, some_trait.name.span())
+        {
+            self.push_link(format_link(some_trait.name.to_string(), lsp_location));
+        }
+    }
+
+    fn gather_type_variable_links(&mut self, var: &TypeVariable) {
+        let var = &*var.borrow();
+        match var {
+            TypeBinding::Bound(typ) => {
+                self.gather_type_links(typ);
+            }
+            TypeBinding::Unbound(..) => (),
+        }
+    }
+
+    fn push_link(&mut self, link: String) {
+        if !self.links.contains(&link) {
```

Comment:
Would it make more sense to use a `HashSet` for `links` to avoid this search?

Reply:
I saw this too but figured for the small amount of links within most types it wouldn't make much of a difference

Reply:
I thought about it, but I wanted the links to preserve insertion order, so if you have `Map<Key, Value>` you'd get links for `Map | Key | Value`.

Given that usually we'll end up with just a few links (usually 2 or 3) doing a linear scan is probably good enough. Alternatively we could use a crate for a dictionary that preserves order... but I didn't know if it was worth it.